File: src/CircleAI.Integration.News/MastodonSource.cs
// MastodonSource.cs
//
// (Phase B3) Mastodon public-timeline / hashtag-timeline reader.
// Anonymous access works for public timelines on most instances.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Net.Http;
using System.Text.Json;
using System.Text.RegularExpressions;
using System.Threading;
using System.Threading.Tasks;
using CircleAI.Integration;

namespace CircleAI.Integration.News;

public sealed record MastodonOptions(string Instance, string? Hashtag = null, string? AccessToken = null);

public sealed class MastodonSource : INewsSource
{
    private readonly HttpClient _http;
    private readonly MastodonOptions _opts;

    public MastodonSource(MastodonOptions opts) : this(opts, new HttpClient()) { }
    public MastodonSource(MastodonOptions opts, HttpClient http)
    {
        _opts = opts ?? throw new ArgumentNullException(nameof(opts));
        _http = http ?? throw new ArgumentNullException(nameof(http));
    }

    public string SourceId   => string.IsNullOrEmpty(_opts.Hashtag)
                                    ? $"mastodon:{_opts.Instance}:public"
                                    : $"mastodon:{_opts.Instance}:#{_opts.Hashtag}";
    public bool   IsConfigured => !string.IsNullOrWhiteSpace(_opts.Instance);

    public async ValueTask<IReadOnlyList<NewsItem>> FetchLatestAsync(int max, CancellationToken ct = default)
    {
        if (max <= 0) throw new ArgumentOutOfRangeException(nameof(max));
        var path = string.IsNullOrEmpty(_opts.Hashtag)
            ? $"/api/v1/timelines/public?limit={Math.Min(max, 40)}"
            : $"/api/v1/timelines/tag/{Uri.EscapeDataString(_opts.Hashtag)}?limit={Math.Min(max, 40)}";
        using var req = new HttpRequestMessage(HttpMethod.Get, _opts.Instance.TrimEnd('/') + path);
        if (!string.IsNullOrWhiteSpace(_opts.AccessToken))
            req.Headers.Authorization = new System.Net.Http.Headers.AuthenticationHeaderValue("Bearer", _opts.AccessToken);
        req.Headers.UserAgent.ParseAdd("CircleAI/1.0 (MastodonSource)");
        using var resp = await _http.SendAsync(req, ct).ConfigureAwait(false);
        resp.EnsureSuccessStatusCode();
        using var doc = await JsonDocument.ParseAsync(
            await resp.Content.ReadAsStreamAsync(ct).ConfigureAwait(false),
            cancellationToken: ct).ConfigureAwait(false);

        if (doc.RootElement.ValueKind != JsonValueKind.Array) return Array.Empty<NewsItem>();
        var list = new List<NewsItem>();
        foreach (var s in doc.RootElement.EnumerateArray())
        {
            var url    = s.TryGetProperty("url",     out var u) ? u.GetString() ?? "" : "";
            var contentHtml = s.TryGetProperty("content", out var c) ? c.GetString() ?? "" : "";
            var pub    = s.TryGetProperty("created_at", out var p) ? p.GetString() : null;
            var tags   = new List<string>();
            if (s.TryGetProperty("tags", out var tagsArr) && tagsArr.ValueKind == JsonValueKind.Array)
                foreach (var tg in tagsArr.EnumerateArray())
                    if (tg.TryGetProperty("name", out var tn)) tags.Add(tn.GetString() ?? "");
            var acct   = s.TryGetProperty("account", out var a) && a.TryGetProperty("acct", out var ac) ? ac.GetString() : null;
            var text   = Regex.Replace(contentHtml, "<[^>]+>", " ").Trim();
            list.Add(new NewsItem(
                ItemId:       url,
                SourceId:     acct ?? SourceId,
                Title:        text.Length > 80 ? text[..80] + "…" : text,
                Summary:      text,
                Url:          Uri.TryCreate(url, UriKind.Absolute, out var ux) ? ux : new Uri("about:blank"),
                PublishedUtc: DateTimeOffset.TryParse(pub, CultureInfo.InvariantCulture, DateTimeStyles.AssumeUniversal, out var dto)
                                ? dto.ToUniversalTime() : DateTimeOffset.MinValue,
                Tags:         tags));
        }
        return list;
    }
}

File: python/src/circle_ai/integration_news/mastodon_source.py
# mastodon_source.py
#
# Port of CircleAI.Integration.News/MastodonSource.cs (C# — the EXACT spec).
#
# (Phase B3) Mastodon public-timeline / hashtag-timeline reader. Anonymous
# access works for public timelines on most instances.
#
# The C# takes an injected ``HttpClient``; the Python port takes an injected
# :class:`IHttpFetcher` and parses the identical JSON array. HTML is stripped
# with the same ``<[^>]+>`` -> " " regex then trimmed.

from __future__ import annotations

import re
from dataclasses import dataclass
from typing import List, Optional
from urllib.parse import quote, urlparse

from circle_ai.integration._util import parse_utc
from circle_ai.integration.contracts import DATETIME_MIN, INewsSource, NewsItem
from circle_ai.integration.http import HttpRequest, IHttpFetcher

_TAG_RX = re.compile(r"<[^>]+>")


def _strip_html(html: str) -> str:
    return _TAG_RX.sub(" ", html).strip()


def _try_absolute(url: str) -> str:
    """Mirror ``Uri.TryCreate(url, UriKind.Absolute)`` — keep an absolute URL,
    else fall back to ``about:blank``.
    """
    if url:
        parsed = urlparse(url)
        if parsed.scheme and parsed.netloc:
            return url
    return "about:blank"


@dataclass(frozen=True, slots=True)
class MastodonOptions:
    """Mirrors ``CircleAI.Integration.News.MastodonOptions`` — ``record(string
    Instance, string? Hashtag = null, string? AccessToken = null)``.
    """

    instance: str
    hashtag: Optional[str] = None
    access_token: Optional[str] = None


class MastodonSource(INewsSource):
    """Port of ``CircleAI.Integration.News.MastodonSource``."""

    def __init__(self, opts: MastodonOptions, http: IHttpFetcher) -> None:
        if opts is None:
            raise ValueError("opts must not be None")
        if http is None:
            raise ValueError("http must not be None")
        self._opts = opts
        self._http = http

    @property
    def source_id(self) -> str:
        if not self._opts.hashtag:
            return f"mastodon:{self._opts.instance}:public"
        return f"mastodon:{self._opts.instance}:#{self._opts.hashtag}"

    @property
    def is_configured(self) -> bool:
        return bool(self._opts.instance and self._opts.instance.strip())

    async def fetch_latest_async(self, max: int) -> List[NewsItem]:
        if max <= 0:
            raise ValueError("max must be positive")
        if not self._opts.hashtag:
            path = f"/api/v1/timelines/public?limit={min(max, 40)}"
        else:
            path = (
                f"/api/v1/timelines/tag/{quote(self._opts.hashtag, safe='')}"
                f"?limit={min(max, 40)}"
            )
        headers = {"User-Agent": "CircleAI/1.0 (MastodonSource)"}
        if self._opts.access_token and self._opts.access_token.strip():
            headers["Authorization"] = f"Bearer {self._opts.access_token}"
        url = self._opts.instance.rstrip("/") + path
        resp = (
            await self._http.send_async(HttpRequest("GET", url, headers))
        ).ensure_success()
        root = resp.json()

        if not isinstance(root, list):
            return []
        result: List[NewsItem] = []
        for s in root:
            if not isinstance(s, dict):
                continue
            url_s = s.get("url") or ""
            content_html = s.get("content") or ""
            pub = s.get("created_at")
            tags: List[str] = []
            tags_arr = s.get("tags")
            if isinstance(tags_arr, list):
                for tg in tags_arr:
                    if isinstance(tg, dict) and "name" in tg:
                        tags.append(tg.get("name") or "")
            acct = None
            a = s.get("account")
            if isinstance(a, dict):
                acct = a.get("acct")
            text = _strip_html(content_html)
            result.append(
                NewsItem(
                    item_id=url_s,
                    source_id=acct if acct else self.source_id,
                    title=(text[:80] + "…") if len(text) > 80 else text,
                    summary=text,
                    url=_try_absolute(url_s),
                    published_utc=parse_utc(pub) if pub is not None else DATETIME_MIN,
                    tags=tags,
                )
            )
        return result
